sdk/trace/trace_test.go

```diff
@@ -1594,6 +1594,46 @@ func TestAddLinksWithMoreAttributesThanLimit(t *testing.T) {
 	}
 }
 
+type frozenClock struct {
+	now time.Time
+}
+type frozenStopwatch struct {
+	started time.Time
+}
+
+func (f frozenStopwatch) Started() time.Time {
+	return f.started
+}
+func (f frozenStopwatch) Stop() time.Duration {
+	return 0
+}
+
+// newFrozenClock returns a clock which stops at time t
+func newFrozenClock(t time.Time) frozenClock {
+	return frozenClock{
+		now: t,
+	}
+}
+
+func (c frozenClock) Stopwatch() Stopwatch {
+	return frozenStopwatch{
+		started: c.now,
+	}
+}
+
+func TestCustomClock(t *testing.T) {
+	te := NewTestExporter()
+	now := time.Now()
+	tp := NewTracerProvider(WithSyncer(te), WithClock(newFrozenClock(now)))
+	tracer := tp.Tracer("custom-clock")
+	_, span := tracer.Start(context.Background(), "test-frozen-clock")
+	time.Sleep(time.Microsecond * 2)
+	span.End()
+	got := te.Spans()[0]
+	assert.Equal(t, now, got.StartTime())
+	assert.Equal(t, now, got.EndTime())
```

Comment:
nit

Here is what I am proposing

```go
	te := NewTestExporter()
	now := time.Now()
	tp := NewTracerProvider(WithSyncer(te), WithClock(newFrozenClock(now)))
	tracer := tp.Tracer("custom-clock")

	_, span := tracer.Start(context.Background(), "test-frozen-clock")
	time.Sleep(time.Microsecond * 2)
	got, err := endSpan(te, span)
	require.NoError(t, err, "should end the span")

	assert.Equal(t, now, got.StartTime(), "StartTime should return the frozen time")
	assert.Equal(t, now, got.EndTime(), "EndTime should return the frozen time")
```

1. a little refactoring
2. `got := te.Spans()[0]` can panic. I see that there is an `endSpan` utility function


I have not tested the proposed changes but I hope it will just work 😉

Reply:
The `checkTime` in `endSpan` would clears the time, making the comparison fail though.